lib/excon/socket.rb

```diff
@@ -364,6 +364,9 @@ def select_with_timeout(socket, type)
       if @data.include?(:deadline)
         request_timeout = request_time_remaining
 
+        # If request timeout is already reached, there's no need to proceed.
+        raise(Excon::Errors::Timeout.by_type(type, 'request')) if request_timeout <= 0
```

Comment:
The only concern is that if we do that, we lose the benefit of being able to tell which specific timeout caused the overall request timeout. It would be powerful if I could set timeout on the entire request, but retry only on connect timeouts.

That's why I opted to raise specific timeout even for requests.

Another alternative is to repeat the whole hierarchy for Request:

```
Timeout
|_ReadTimeout
|_WriteTimeout
|_ConnectTimeout
  |_ConnectReadTimeout
  |_ConnectWriteTimeout
|_RequestTimeout
  |_RequestReadTimeout
  |_RequestWriteTimeout
  |_RequestConnectTimeout
    |_RequestConnectReadTimeout
    |_RequestConnectWriteTimeout
```

this seems a little excessive however.

Are you sure you'd want to hide the specific timeout error behind RequestTimeout?

Reply:
Forgot to mention this detail: there is already an exception called `RequestTimeout` (code 408), which would be in conflict if we did this.

Reply:
Oops, right...

For some context, if I recall correctly the request timeout here is harder to pin down to a particular cause in the same way.

ie I expect read timeout is specifically that I started waiting at the beginning of a read and it took longer than I wanted, so I raised. Same thing for write or connect errors.

I believe the intention with the error we are talking about here is actually a timeout for the entirety of the request. So the timer starts as soon as the request process begins and it times out as soon as the limit is surpassed. That being said, it's possible for instance that the timeout actually would occur in between the other activities during excon's own processing. ie if connect took just slightly less than the request limit, the timeout would technically occur in between connect and write. In that case, it wouldn't really be correct to say that the timout was either connect or write (and hence the generic error we see here).

Does that make sense?

All that being said, I'm not sure that "request" is a very good descriptor of these expectations. Maybe it should be something like "OverallTimeout" or "TotalTimeout" or something (I'm definitely open to suggestions). It's also possible that this is more trouble than it's worth and something like what you have done is good enough. What do you think?

I'm glad to happily surprise you in terms of responsiveness, though paradoxically I was feeling bad for how long it took to get back to you...

Reply:
> It's also possible that this is more trouble than it's worth and something like what you have done is good enough.

Could be, but you're right that we have to think this through.

> it's possible for instance that the timeout actually would occur in between the other activities

It's true, we only check monotonic clock once per operation, so we can't know when the deadline was actually crossed. Could've been during the beginning of the current phase, the end of the previous one, or in between.

But if you think about it, for "RequestTimeout" we don't actually care exactly when timeout occurred. We only care which phases of request completed successfully at the moment of raising this error. So maybe we indeed should go with something like `TotalTimeout` (better name pending), but add a piece of metadata on it, like:

```ruby
rescue Excon::Errors::TotalTimeout => e
  puts e.completed_request_phases # => [:connect_write, :connect_read, :write]
end
```

If request timeout is not set by the user, then we fallback to specific timeout errors. I don't know the code deep enough to tell if this is easy to do, but at least it avoids the need for impossible timeout precision. How does this sound?

Reply:
Sounds good to me. Naming wise maybe we could combine the old and new ideas, so something like `OverallRequestTimeout` or `TotalRequestTimeout`? I kind of like overall between those I think. What do you think?

Reply:
`OverallRequestTimeout` looks like a good option. We could also consider a different angle: `RequestDurationTimeout`, `RequestSpanTimeout`. What do you think about these?

Reply:
Oh, yeah. I really like `RequestDurationTimeout`. I think that is probably my new favorite, great idea!